localstack/services/s3/utils.py

```diff
@@ -151,6 +156,92 @@ def verify_checksum(checksum_algorithm: str, data: bytes, request: Dict):
         )
 
 
+def decode_aws_chunked_object(
+    stream: IO[bytes],
+    buffer: IO[bytes],
+    content_length: int,
+    checksum_algorithm: str = None,
+    checksum_value: str = None,
+) -> tuple[str, str]:
+    """
+
+    :param stream: the original stream to read
+    :param buffer: the stream where we set the decoded data
+    :param content_length: the total length of the original stream
+    :param checksum_algorithm:
+    :param checksum_value:
+    :return:
+    """
+    buffer.seek(0)
+    buffer.truncate()
+    # We have 2 cases:
+    # The client gave a checksum value, we will need to compute the value and validate it against
+    # or the client have an algorithm value only, and we need to compute the checksum
+    checksum = None
+    calculated_checksum = None
+    if checksum_algorithm:
+        checksum = get_s3_checksum(checksum_algorithm)
+    etag = hashlib.md5(usedforsecurity=False)
+
+    written = 0
+    while written < content_length:
+        line = stream.readline()
+        chunk_length = int(line.split(b";")[0], 16)
+
+        while chunk_length > 0:
+            amount = min(chunk_length, S3_CHUNK_SIZE)
+            data = stream.read(amount)
+            buffer.write(data)
+
+            real_amount = len(data)
+            chunk_length -= real_amount
+            written += real_amount
+
+            if checksum_algorithm:
+                checksum.update(data)
+            etag.update(data)
+
+        # remove trailing \r\n
+        stream.read(2)
+
+    trailing_headers = []
+    next_line = stream.readline()
+
+    if next_line:
+        try:
+            chunk_length = int(next_line.split(b";")[0], 16)
+            if chunk_length != 0:
+                LOG.warning("The S3 object body didn't conform to the aws-chunk format")
+        except ValueError:
+            trailing_headers.append(next_line.strip())
+
+        # try for trailing headers after
+        while line := stream.readline():
+            trailing_header = line.strip()
+            if trailing_header:
+                trailing_headers.append(trailing_header)
+
+    # look for the checksum header in the trailing headers
+    # TODO: we could get the header key from x-amz-trailer as well
```

Comment:
I think it would make sense to read the `x-amz-trailer` header, because the hard-coded `x-amz-checksum-*` seems a bit dangerous.

Reply:
I'll keep this for the streaming provider/new decoder! removed now from this PR